### PR TITLE
fix(sdk)!: wrong order of objects returned by Drive

### PR DESCRIPTION
**Pull request by lklimek**
## Issue being fixed or feature implemented

Rust Dash SDK does not preserve order of items returned by Drive.

## What was done?

Replaced BTreeMap with IndexMap.

## How Has This Been Tested?

GHA

## Breaking Changes

In Rust SDK, returned objects are IndexMap/IndexSet instead of previous BTreeMap.

### Affected types
* RetrievedObjects, and all types that use it
* RetrievedIntegerValue renamed to RetrievedValues
* DataContractHistory
* ProposerBlockCounts

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

## Summary by CodeRabbit

- **New Features**
	- Introduced `IndexMap` for improved collection handling in various components.
	- Enhanced `MockResponse` trait to support serialization/deserialization for `IndexMap`.
	- Added new methods in `MockContextProvider` for better data contract management and quorum key retrieval.

- **Bug Fixes**
	- Improved error handling in multiple test cases, ensuring robustness against invalid inputs.

- **Documentation**
	- Updated comments for clarity and consistency across methods and test cases.

- **Tests**
	- Expanded test coverage for contested resources and data contracts, including new test scenarios and refined assertions.
	- Updated existing test cases to improve result handling and assertions.

### CODE REVIEW

**Comment by lklimek** [comment]
@coderabbitai review

File: packages/rs-sdk/tests/fetch/document.rs [hunk 1]
```diff
@@ -34,7 +34,8 @@ async fn document_read() {
     let first_doc = Document::fetch_many(&sdk, all_docs_query)
         .await
         .expect("fetch many documents")
-        .pop_first()
+        .into_iter()
+        .next()
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-sdk/tests/fetch/document.rs:38]: These usages of BTreeMap are defined in crates that are out of scope of this change.

File: packages/rs-sdk/src/mock/requests.rs [hunk 2]
```diff
@@ -115,6 +115,29 @@ impl<K: Ord + MockResponse, V: MockResponse> MockResponse for BTreeMap<K, V> {
...
+
+    fn mock_serialize(&self, sdk: &MockDashPlatformSdk) -> Vec<u8> {
+        let data: IndexMap<Vec<u8>, Vec<u8>> = self
+            .iter()
+            .map(|(k, v)| (k.mock_serialize(sdk), v.mock_serialize(sdk)))
+            .collect();
+
+        bincode::serde::encode_to_vec(data, BINCODE_CONFIG).expect("encode BTreeMap")
+    }
+}
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-sdk/src/mock/requests.rs]: IndexMap does not implement bincode::Encode, so we cannot use non-serde functions there.

File: packages/rs-sdk/src/mock/requests.rs [hunk 3]
```diff
@@ -115,6 +115,29 @@ impl<K: Ord + MockResponse, V: MockResponse> MockResponse for BTreeMap<K, V> {
     }
 }
 
+impl<K: Hash + Eq + MockResponse, V: MockResponse> MockResponse for IndexMap<K, V> {
+    fn mock_deserialize(sdk: &MockDashPlatformSdk, buf: &[u8]) -> Self
+    where
+        Self: Sized,
+    {
+        let (data, _): (IndexMap<Vec<u8>, Vec<u8>>, _) =
+            bincode::serde::decode_from_slice(buf, BINCODE_CONFIG).expect("decode BTreeMap");
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-sdk/src/mock/requests.rs]: done

File: packages/rs-drive-proof-verifier/src/proof.rs [hunk 4]
```diff
@@ -1638,23 +1639,25 @@ impl FromProof<platform::GetContestedResourceIdentityVotesRequest> for Vote {
...
                 return Err(Error::ResponseDecodeError {
                     error: format!("expected 1 vote, got {}", v.len()),
                 })
             }
             Some(v) if v.is_empty() => return Ok((None, mtd, proof)),
-            Some(v) => v.pop_first().expect("is_empty() must detect empty map"),
+            Some(v) => v
+                .into_iter()
+                .next()
+                .expect("is_empty() must detect empty map"),
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-drive-proof-verifier/src/proof.rs:1664]: this is correct, this expect() will never be called as above, as in `Some(v) if v.is_empty()` match arm we check if there is some data in the response.

**Comment by lklimek** [comment]
@coderabbitai review

File: packages/rs-drive-proof-verifier/src/types.rs [hunk 5]
```diff
@@ -112,14 +114,14 @@ impl Contenders {
         &self,
         document_type: &DocumentType,
         platform_version: &PlatformVersion,
-    ) -> Result<BTreeMap<Identifier, Contender>, crate::Error> {
+    ) -> Result<IndexMap<Identifier, Contender>, crate::Error> {
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-drive-proof-verifier/src/types.rs:117]: reverted to BTreeMap

**Review by shumkov — APPROVED** [review]

**Pull request merged**